File: java/src/tarehart/rlbot/steps/travel/SlideToPositionStep.java
package tarehart.rlbot.steps.travel;

import tarehart.rlbot.AgentInput;
import tarehart.rlbot.AgentOutput;
import tarehart.rlbot.input.CarData;
import tarehart.rlbot.math.VectorUtil;
import tarehart.rlbot.math.vector.Vector2;
import tarehart.rlbot.planning.Plan;
import tarehart.rlbot.routing.PositionFacing;
import tarehart.rlbot.planning.SteerUtil;
import tarehart.rlbot.steps.Step;

import java.awt.*;
import java.awt.geom.Line2D;
import java.util.Optional;
import java.util.function.Function;

public class SlideToPositionStep implements Step {

    private Function<AgentInput, Optional<PositionFacing>> targetFunction;

    private static final int AIM_AT_TARGET = 0;
    private static final int TRAVEL = 1;
    private static final int SLIDE_SPIN = 2;

    private int phase = AIM_AT_TARGET;

    private Integer turnDirection = null;
    private PositionFacing target;

    private Plan plan;

    public SlideToPositionStep(Function<AgentInput, Optional<PositionFacing>> targetFunction) {
        this.targetFunction = targetFunction;
    }


    @Override
    public Optional<AgentOutput> getOutput(AgentInput input) {

        if (Plan.activePlan(plan).isPresent()) {
            Optional<AgentOutput> output = plan.getOutput(input);
            if (output.isPresent()) {
                return output;
            }
        }

        Optional<PositionFacing> targetOption = targetFunction.apply(input);
        if (!targetOption.isPresent()) {
            return Optional.empty();
        }
        target = targetOption.get();

        CarData car = input.getMyCarData();

        Vector2 toTarget = target.getPosition().minus(car.getPosition().flatten());

        if (phase == AIM_AT_TARGET) {

            if (toTarget.magnitude() < 10) {
                return Optional.empty();
            }


            double angle = Vector2.Companion.angle(toTarget, car.getOrientation().getNoseVector().flatten());
            if (angle < Math.PI / 12) {
                phase = TRAVEL;
            } else {
                return Optional.of(SteerUtil.steerTowardGroundPosition(car, input.getBoostData(), target.getPosition()));
            }
        }

        if (phase == TRAVEL) {

            double distance = toTarget.magnitude();
            double slideDistance = getSlideDistance(car.getVelocity().magnitude());

            if (distance < slideDistance) {
                phase = SLIDE_SPIN;
            } else {

                double correctionRadians = toTarget.correctionAngle(target.getFacing());

                Vector2 offsetVector = VectorUtil.INSTANCE
                        .rotateVector(toTarget, -Math.signum(correctionRadians) * Math.PI / 2)
                        .scaledToMagnitude(10);

                Vector2 waypoint = target.getPosition().plus(offsetVector);

                Optional<Plan> sensibleFlip = SteerUtil.getSensibleFlip(car, waypoint);
                if (sensibleFlip.isPresent()) {
                    this.plan = sensibleFlip.get();
                    return this.plan.getOutput(input);
                }

                return Optional.of(SteerUtil.steerTowardGroundPosition(car, input.getBoostData(), waypoint).withBoost(car.getBoost() > 50));
            }
        }

        if (phase == SLIDE_SPIN) {
            if (turnDirection == null) {
                turnDirection = (int) Math.signum(car.getOrientation().getNoseVector().flatten().correctionAngle(toTarget));
            }

            double correctionRadians = car.getOrientation().getNoseVector().flatten().correctionAngle(target.getFacing());
            double futureRadians = correctionRadians + car.getSpin().getYawRate() * .3;

            if (futureRadians * turnDirection < 0 && Math.abs(futureRadians) < Math.PI / 4) {
                return Optional.empty(); // Done orienting.
            }

            return Optional.of(new AgentOutput().withAcceleration(1).withSteer(-turnDirection).withSlide());
        }

        return Optional.empty(); // Something went wrong
    }

    private double getSlideDistance(double speed) {
        return speed * .8;
    }

    @Override
    public String getSituation() {
        return "Sliding to position";
    }

    @Override
    public boolean canInterrupt() {
        return plan == null || plan.canInterrupt();
    }

    @Override
    public void drawDebugInfo(Graphics2D graphics) {
        if (target != null) {
                graphics.setColor(new Color(75, 214, 214));
                graphics.setStroke(new BasicStroke(1));
                Vector2 position = target.getPosition();
                int crossSize = 2;
                graphics.draw(new Line2D.Double(position.getX() - crossSize, position.getY() - crossSize, position.getX() + crossSize, position.getY() + crossSize));
                graphics.draw(new Line2D.Double(position.getX() - crossSize, position.getY() + crossSize, position.getX() + crossSize, position.getY() - crossSize));
        }
    }
}

File: java/src/tarehart/rlbot/steps/travel/SlideToPositionStep.kt
package tarehart.rlbot.steps.travel

import tarehart.rlbot.AgentInput
import tarehart.rlbot.AgentOutput
import tarehart.rlbot.input.CarData
import tarehart.rlbot.math.VectorUtil
import tarehart.rlbot.math.vector.Vector2
import tarehart.rlbot.planning.Plan
import tarehart.rlbot.routing.PositionFacing
import tarehart.rlbot.planning.SteerUtil
import tarehart.rlbot.steps.NestedPlanStep
import tarehart.rlbot.steps.Step

import java.awt.*
import java.awt.geom.Line2D
import java.util.Optional
import java.util.function.Function

class SlideToPositionStep(private val targetFunction: (AgentInput) -> PositionFacing?) : NestedPlanStep() {

    private var phase = AIM_AT_TARGET

    private var turnDirection: Int? = null

    private var target: PositionFacing? = null

    override fun getLocalSituation(): String {
        return "Sliding to position"
    }

    override fun doComputationInLieuOfPlan(input: AgentInput): Optional<AgentOutput> {

        val targetOption = targetFunction.invoke(input) ?: return Optional.empty()

        target = targetOption

        val car = input.myCarData

        val toTarget = targetOption.position.minus(car.position.flatten())

        if (phase == AIM_AT_TARGET) {

            if (toTarget.magnitude() < 10) {
                return Optional.empty()
            }


            val angle = Vector2.angle(toTarget, car.orientation.noseVector.flatten())
            if (angle < Math.PI / 12) {
                phase = TRAVEL
            } else {
                return Optional.of(SteerUtil.steerTowardGroundPosition(car, input.boostData, targetOption.position))
            }
        }

        if (phase == TRAVEL) {

            val distance = toTarget.magnitude()
            val slideDistance = getSlideDistance(car.velocity.magnitude())

            if (distance < slideDistance) {
                phase = SLIDE_SPIN
            } else {

                val correctionRadians = toTarget.correctionAngle(targetOption.facing)

                val offsetVector = VectorUtil
                        .rotateVector(toTarget, -Math.signum(correctionRadians) * Math.PI / 2)
                        .scaledToMagnitude(10.0)

                val waypoint = targetOption.position.plus(offsetVector)

                val sensibleFlip = SteerUtil.getSensibleFlip(car, waypoint)
                if (sensibleFlip.isPresent) {
                    return startPlan(sensibleFlip.get(), input)
                }

                return Optional.of(SteerUtil.steerTowardGroundPosition(car, input.boostData, waypoint).withBoost(car.boost > 50))
            }
        }

        if (phase == SLIDE_SPIN) {
            val turnDir = turnDirection ?: Math.signum(car.orientation.noseVector.flatten().correctionAngle(toTarget)).toInt()
            turnDirection = turnDir

            val correctionRadians = car.orientation.noseVector.flatten().correctionAngle(targetOption.facing)
            val futureRadians = correctionRadians + car.spin.yawRate * .3

            return if (futureRadians * turnDir < 0 && Math.abs(futureRadians) < Math.PI / 4) {
                Optional.empty() // Done orienting.
            } else Optional.of(AgentOutput().withAcceleration(1.0).withSteer(-turnDir.toDouble()).withSlide())

        }

        return Optional.empty() // Something went wrong
    }

    private fun getSlideDistance(speed: Double): Double {
        return speed * .8
    }

    override fun drawDebugInfo(graphics: Graphics2D) {
        super.drawDebugInfo(graphics)

        target?.let {
            graphics.color = Color(75, 214, 214)
            graphics.stroke = BasicStroke(1f)
            val position = it.position
            val crossSize = 2
            graphics.draw(Line2D.Double(position.x - crossSize, position.y - crossSize, position.x + crossSize, position.y + crossSize))
            graphics.draw(Line2D.Double(position.x - crossSize, position.y + crossSize, position.x + crossSize, position.y - crossSize))
        }
    }

    companion object {

        private val AIM_AT_TARGET = 0
        private val TRAVEL = 1
        private val SLIDE_SPIN = 2
    }
}
